Change databricks build to dynamically create a cluster

tgravescs opened the pull request:
This changes it so we dynamically create a new databricks cluster everytime we kick off a nightly build.  It deletes the cluster at the end of the run as well.  This is the basic functionality that we have now and we can continue enhance it later.  
<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

tgravescs commented:
build

tgravescs commented:
tested here: https://ci-dev.ngcc.nvidia.com/job/spark/job/tgraves-test-databricks-create/

tgravescs commented:
@NvTimLiu

jlowe commented on jenkins/Jenkinsfile.databricksnightly:
```diff
@@ -78,7 +80,7 @@ pipeline {
                         sh "rm -rf spark-rapids-ci.tgz"
                         sh "tar -zcvf spark-rapids-ci.tgz *"
-                        sh "python3.6 ./jenkins/databricks/run-tests.py -c $CLUSTER_ID -z ./spark-rapids-ci.tgz -t $DATABRICKS_TOKEN -p /home/svcngcc/.ssh/id_rsa -l ./jenkins/databricks/build.sh -j $CI_RAPIDS_JAR -b $DATABRICKS_VERSION -k $SPARK_VERSION -a $SCALA_VERSION -f $CUDF_VERSION -u $CUDA_VERSION -m $CI_CUDF_JAR"
+                        sh "python3.6 ./jenkins/databricks/run-tests.py -z ./spark-rapids-ci.tgz -t $DATABRICKS_TOKEN -p /home/svcngcc/.ssh/id_rsa -l ./jenkins/databricks/build.sh -j $CI_RAPIDS_JAR -b $DATABRICKS_VERSION -k $SPARK_VERSION_TO_INSTALL_JARS -a $SCALA_VERSION -f $CUDF_VERSION -u $CUDA_VERSION -m $CI_CUDF_JAR -w $PUBLIC_KEY -o $CLUSTER_ID_LOC"
```
Seems to me the output file is needed because we're overloaading run-tests to not only run tests on a cluster but also create the cluster.  Would it make more sense to have the Jenkins setup explicitly create the cluster first with a cluster creation script that returns its cluster ID as the lone thing to stdout (all other logging goes to stderr), and Jenkins tracks that cluster Id in a variable?  Admittedly, I'm not enough of a Jenkins expert to know if this makes sense.

tgravescs replied:
Yeah I think we can break it up that way and use env variable to pass between to get to the post step where it shuts down. I had a bit of problems with that but I think I read how to do it properly so I'll try it out.

revans2 commented on jenkins/databricks/clusterutils.py:
```diff
@@ -0,0 +1,149 @@
+
+    @staticmethod
+    def generate_create_templ(sshKey, initScriptFile, version):
+        # TODO (in followup) - create directory if not there
```
Is there a follow on issue for this?  Or do you plan to tackle it when/if it becomes a problem?

tgravescs replied:
this was for enhancing to do dev builds and such. Or if we add more testing envs.

tgravescs replied:
I'll just remove it for now

revans2 commented on jenkins/databricks/clusterutils.py:
```diff
@@ -0,0 +1,149 @@
+        # TODO (in followup) - create directory if not there
+        init_script_loc = 'dbfs:/databricks/ci_init_scripts/' + initScriptFile
+        dt = datetime.now()
+        databricks_runtime_version = '7.0.x-gpu-ml-scala2.12'
```
nit: Should this be configurable because we are going to support multiple versions of data bricks? We can do it later, just curious because it looks like it would be a small change to make it work.

tgravescs replied:
yep, agree, that is why I pulled to variable here, but plan on doing later when needed.

revans2 commented on jenkins/databricks/clusterutils.py:
```diff
@@ -0,0 +1,149 @@
+        init_script_loc = 'dbfs:/databricks/ci_init_scripts/' + initScriptFile
+        dt = datetime.now()
+        databricks_runtime_version = '7.0.x-gpu-ml-scala2.12'
+        idle_timeout = 240
```
nit: Might be good to have a follow on so that it is tied to the timeout for the Jenkins job so we don't have a config for one timeout and a separate hard coded timeout.  I really hope we never get over 4 hours to run the tests, but it could also save us some money in the case of an error.

tgravescs replied:
I'll just make it a parameter now

revans2 reviewed (commented):
Just some nits that really don't matter much at all.  Happy to approve it once the comment from Jason is addressed

tgravescs commented:
added a create.py script and split create off from run-tests.py.  Added a bunch more options to make it more configurable.  Passing clusterid back in stdout from create script and then pass it in jenkinsfile via environment variable.

tgravescs commented:
running one file test here: https://ci-dev.ngcc.nvidia.com/job/spark/job/tgraves-test-databricks-create/36/

tgravescs commented:
build

jlowe commented on jenkins/databricks/create.py:
```diff
@@ -0,0 +1,96 @@
+  print('-e is ' + str(num_workers), file=sys.stderr)
+
+  if not sshkey:
+      print("You must specify an sshkey!")
```
This should go to stderr

jlowe commented on jenkins/databricks/create.py:
```diff
@@ -0,0 +1,96 @@
+      sys.exit(2)
+
+  if not token:
+      print("You must specify an token!")
```
This should go to stderr

jlowe commented on jenkins/databricks/run-tests.py:
```diff
@@ -18,65 +18,49 @@
   except getopt.GetoptError:
       print(
-          'run-tests.py -s <workspace> -t <token> -c <clusterid> -p <privatekeyfile> -n <skipstartingcluster> -l <localscript> -d <scriptdestinatino> -z <sparktgz> -j <cirapidsjar> -b <databricksversion> -k <sparkversion> -a <scalaversion> -f <cudfversion> -u <cudaversion> -m <cicudfjar> -v <basesparkpomversion>')
+          'run-tests.py -s <workspace> -t <token> -c <clusterid> -p <privatekeyfile> -l <localscript> -d <scriptdestinatino> -z <sparktgz> -j <cirapidsjar> -b <databricksversion> -k <sparkversion> -a <scalaversion> -f <cudfversion> -u <cudaversion> -m <cicudfjar> -v <basesparkpomversion>')
```
This should go to stderr, but there are a number of other pre-existing prints here that should as well.  Totally OK if this is a low-priority cleanup issue to followup.

tgravescs replied:
yeah I didn't change the run-tests.py to go to stderr by default as it wasn't required at this point.  as long as you are ok I'd like to do it in followup when we split that up more

tgravescs commented:
build

tgravescs commented:
3.1.0 failure we have pr up for

tgravescs commented:
build

tgravescs commented:
same 3.1.0 failure with ParquetRowConverter that should have been fixed second time I built. will try again

tgravescs commented:
build

jlowe commented:
> same 3.1.0 failure with ParquetRowConverter that should have been fixed second time I built. will try again

The PR needs to be upmerged with latest on branch-0.3 to pick up the fix.

tgravescs commented:
build
